File: Stacks_Queues/go/StackMin.go
package main

import (
  "fmt"
  "types"
)


// Stack implementation
type StackMin struct {
  elements []int
  mins types.DoublyLinkedList
  top int
}


func (s *StackMin) Push(x int) {
  s.elements = append(s.elements, x)
  s.top = x

  if s.mins.Head == nil {
    s.mins.Append(x)
    return
  }

  if x <= s.mins.Tail.Data {
    s.mins.Append(x)
  }
}


func (s *StackMin) Pop() (int, bool) {
  top := s.top
  length := len(s.elements)

  if length == 0 {
    return 0, false
  }

  s.elements = s.elements[:length-1]

  if top == s.mins.Tail.Data {
    s.mins.DeleteLast()
  }

  if length != 1 {
    s.top = s.elements[length-2]
  }

  return top, true
}


func PrintStack(s StackMin) {
  fmt.Println(s.elements)
}


func (s *StackMin) Min() int {
  return s.mins.Tail.Data
}


func main() {
  var s StackMin
  s.Push(34)
  s.Push(2)
  s.Push(32)
  s.Push(-3)
  s.Pop()
  s.Pop()

  PrintStack(s)
  fmt.Println(s.Min())
}

File: Stacks_Queues/python/StackMin.py
from DoublyLinkedList import DoublyLinkedList


class StackMin(object):
    def __init__(self):
        self.elements = []
        self.mins = DoublyLinkedList()
        self.top = None

    def push(self, x):
        self.elements.append(x)
        self.top = x

        if self.mins.head == None:
            self.mins.append(x)
            return

        if x <= self.mins.tail.data:
            self.mins.append(x)

    def pop(self):
        top = self.top
        length = len(self.elements)

        if not length:
            return

        self.elements.pop()

        if top == self.mins.tail.data:
            self.mins.delete_last()

        if length != 1:
            self.top = self.elements[length-2]

        return top

    def min(self):
        return self.mins.tail.data

def main():
    s = StackMin()
    s.push(34)
    s.push(2)
    s.push(32)
    s.push(-3)
    s.pop()
    s.pop()

    print(s.elements)
    print(s.min())


if __name__ == '__main__':
    main()
